pkg/tms/tmsUtils.go

```diff
@@ -0,0 +1,358 @@
+	if err != nil {
+		log.Entry().WithError(err).Fatal("Failed to prepare client for talking with TMS")
+	}
+	return commuInstance
+}
+
+func UploadDescriptors(config Options, communicationInstance CommunicationInterface, utils TmsUtils) error {
+	description := config.CustomDescription
+	namedUser := config.NamedUser
+	nodeName := config.NodeName
+	mtaVersion := config.MtaVersion
+	nodeNameExtDescriptorMapping := config.NodeExtDescriptorMapping
+	mtaPath := config.MtaPath
+
+	if config.Verbose {
+		log.Entry().Info("The step will use the following values:")
+		log.Entry().Infof("- description: %v", config.CustomDescription)
+
+		if len(nodeNameExtDescriptorMapping) > 0 {
+			log.Entry().Infof("- mapping between node names and MTA extension descriptor file paths: %v", nodeNameExtDescriptorMapping)
+		}
+		log.Entry().Infof("- MTA path: %v", mtaPath)
+		log.Entry().Infof("- MTA version: %v", mtaVersion)
+		if namedUser != "" {
+			log.Entry().Infof("- named user: %v", namedUser)
+		}
+		log.Entry().Infof("- node name: %v", nodeName)
+	}
+
+	if len(nodeNameExtDescriptorMapping) > 0 {
+		nodes, errGetNodes := communicationInstance.GetNodes()
+		if errGetNodes != nil {
+			log.SetErrorCategory(log.ErrorService)
+			return fmt.Errorf("failed to get nodes: %w", errGetNodes)
+		}
+
+		mtaYamlMap, errGetMtaYamlAsMap := GetYamlAsMap(utils, "mta.yaml")
+		if errGetMtaYamlAsMap != nil {
+			log.SetErrorCategory(log.ErrorConfiguration)
+			return fmt.Errorf("failed to get mta.yaml as map: %w", errGetMtaYamlAsMap)
+		}
+		_, isIdParameterInMap := mtaYamlMap["ID"]
+		_, isVersionParameterInMap := mtaYamlMap["version"]
+		if !isIdParameterInMap || !isVersionParameterInMap {
+			var errorMessage string
+			if !isIdParameterInMap {
+				errorMessage += "parameter 'ID' is not found in mta.yaml\n"
+			}
+			if !isVersionParameterInMap {
+				errorMessage += "parameter 'version' is not found in mta.yaml\n"
+			}
+			log.SetErrorCategory(log.ErrorConfiguration)
+			return errors.New(errorMessage)
+		}
+
+		// validate the whole mapping and then throw errors together, so that user can get them after a single pipeline run
+		nodeIdExtDescriptorMapping, errGetNodeIdExtDescriptorMapping := FormNodeIdExtDescriptorMappingWithValidation(utils, nodeNameExtDescriptorMapping, nodes, mtaYamlMap, mtaVersion)
+		if errGetNodeIdExtDescriptorMapping != nil {
+			log.SetErrorCategory(log.ErrorConfiguration)
+			return errGetNodeIdExtDescriptorMapping
+		}
+
+		for nodeId, mtaExtDescriptorPath := range nodeIdExtDescriptorMapping {
+			obtainedMtaExtDescriptor, errGetMtaExtDescriptor := communicationInstance.GetMtaExtDescriptor(nodeId, fmt.Sprintf("%v", mtaYamlMap["ID"]), mtaVersion)
+			if errGetMtaExtDescriptor != nil {
+				log.SetErrorCategory(log.ErrorService)
+				return fmt.Errorf("failed to get MTA extension descriptor: %w", errGetMtaExtDescriptor)
+			}
+
+			if obtainedMtaExtDescriptor != (MtaExtDescriptor{}) {
```

Comment:
How about checking for the id here instead of checking if it is not empty. Not being empty strictly speaking does not guarantee you having an Id later, which is what you need for `UpdateMtaExtDescriptor`.

Reply:
We rely on TMS API: if there is a response then ID is not empty. 
The check empty - not empty is required for the logic  (if MtaExtDescriptor exists then update it, if not - upload it)

Reply:
ok, got it